### PR TITLE
Added abstract wrapper class (resolves #13)

### LINKED ISSUE #13: Add PipelineWrapperBuilder to toil-lib
PipelineWrapperBuilder is a new class used by pipeline wrapper.py files to implement an entry point for docker containers.

### PR DESCRIPTION
**Pull request by JakeNarkizian**
This class will be used by wrapper.py files in dockerized workflows to implement an entry point.

### CODE REVIEW

**Comment by JakeNarkizian** [comment]
This PR is not ready for review. @alex-hancock please add comments to this PR with any bugs you find during testing.

**Comment by JakeNarkizian** [comment]
@jvivian

**Comment by jvivian** [comment]
@JakeNarkizian — I'll review this tomorrow!

File: pipelineWrapper.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+class UserError(Exception):
+    pass
+
+
+class DefinitionError(Exception):
+    pass
+
+
+def require(expression, message):
```
Comment (jvivian) [inline pipelineWrapper.py]: `require` is already in Toil-lib, just import it:   `from toil_lib import require`

File: src/toil_lib/pipelineWrapperBuilder.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                                   self._mount])
+            if self._no_clean and args.no_clean:
+                log.info('Flag "--no-clean" was used, therefore %s was not deleted.', self._workdir)
+            else:
+                log.info('Cleaning up temporary directory: %s', self._workdir)
+                shutil.rmtree(self._workdir)
+
+    def get_args(self):
+        """
+        Use this context manager to add arguments to an argparse object with the add_argument
```
Comment (jvivian) [inline src/toil_lib/pipelineWrapperBuilder.py]: Is this still a context manager?  I thought a contextmanager needed a `@contextmanager` decorator or needed to be a class with `__enter__` and  `__exit__` method. 

Can you provide an example of how someone would add an argument via this method?

File: src/toil_lib/pipelineWrapperBuilder.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+import argparse
+import json
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import textwrap
+from contextlib import contextmanager
```
Comment (jvivian) [inline src/toil_lib/pipelineWrapperBuilder.py]: unused import

File: src/toil_lib/pipelineWrapperBuilder.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                                help='If this flag is used, temporary work directory is not '
+                                     'cleaned.')
+        if self._resume:
+            parser.add_argument('--resume', action='store_true',
+                                help='If this flag is used, a previously uncleaned workflow in the'
+                                     ' same directory will be resumed')
+        return parser
+
+    def _prepare_mount(self, args):
+        assert args is not None
```
Comment (jvivian) [inline src/toil_lib/pipelineWrapperBuilder.py]: `assert args is not None, 'No arguments provided, use the get_args method first.'`

File: src/toil_lib/pipelineWrapperBuilder.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+class UserError(Exception):
+    pass
+
+
+class DefinitionError(Exception):
+    pass
+
+
+def require(expression, message):
```
Comment (jvivian) [inline src/toil_lib/pipelineWrapperBuilder.py]: Import this from toil-lib, will also let you remove `UserError`

File: src/toil_lib/pipelineWrapperBuilder.py (hunk 2 above)
Comment (JakeNarkizian) [inline src/toil_lib/pipelineWrapperBuilder.py]: This is no longer a context manager. It returns a parser object that a wrapper script adds arguments to and parses. Here is a simple example:

```
parser = pipeline.get_args()
parser.add_arguement(foo bar baz qux)
```

File: src/toil_lib/pipelineWrapperBuilder.py (hunk 4 above)
Comment (JakeNarkizian) [inline src/toil_lib/pipelineWrapperBuilder.py]: I'm removing this assert entirely and args as a param of this method, its not actually used anywhere other than the assert.

File: src/toil_lib/pipelineWrapperBuilder.py (hunk 2 above)
Comment (jvivian) [inline src/toil_lib/pipelineWrapperBuilder.py]: Correct docstring to remove context manager part.

**Comment by JakeNarkizian** [comment]
@jvivian

**Comment by jvivian** [comment]
Thanks @JakeNarkizian — LGTM.  PR contains code additions only.

**Comment by jvivian** [comment]
@JakeNarkizian — Can you rebase onto master?

File: pipelineWrapper.py (hunk 1 above)
Comment (hannes-ucsc) [inline pipelineWrapper.py]: Require is in bd2k-python-lib. Are you suggesting a re-export, @jvivian?

File: src/toil_lib/pipelineWrapperBuilder.py (hunk 5 above)
Comment (hannes-ucsc) [inline src/toil_lib/pipelineWrapperBuilder.py]: See above

**Comment by hannes-ucsc** [comment]
@JakeNarkizian, do you plan on incorporating the shared wrapper in the rnaseq-cgl-pipeline, too?

**Comment by JakeNarkizian** [comment]
@hannes-ucsc This is ready for review. Also to answer your question: yes. I have a wrapper.py file which subclasses this builder class. I will open a PR adding this to the rnaseq repo as soon as this pr is merged.

**Comment by JakeNarkizian** [comment]
@hannes-ucsc this is ready to be merged.

**Pull request merged**